consensus/src/quorum_store/proof_coordinator.rs

```diff
@@ -64,40 +71,78 @@ impl IncrementalProofState {
             return Err(SignedDigestError::DuplicatedSignature);
         }
 
-        self.aggregated_signature
-            .insert(signed_digest.signer(), signed_digest.signature());
+        match validator_verifier.get_voting_power(&signed_digest.signer()) {
+            Some(voting_power) => {
+                let signer = signed_digest.signer();
+                if self
+                    .aggregated_signature
+                    .insert(signer, signed_digest.signature())
+                    .is_none()
+                {
+                    self.aggregated_voting_power += voting_power as u128;
+                } else {
+                    error!(
+                        "Author already in aggregated_signatures right after rechecking: {}",
+                        signer
+                    );
+                }
+            },
+            None => {
+                error!(
+                    "Received signature from author not in validator set: {}",
+                    signed_digest.signer()
+                );
+                return Err(SignedDigestError::InvalidAuthor);
+            },
+        }
+
         Ok(())
     }
 
     fn ready(&self, validator_verifier: &ValidatorVerifier, my_peer_id: PeerId) -> bool {
-        self.aggregated_signature.contains_key(&my_peer_id)
-            && validator_verifier
-                .check_voting_power(self.aggregated_signature.keys())
-                .is_ok()
+        if self.aggregated_signature.contains_key(&my_peer_id)
+            && self.aggregated_voting_power >= validator_verifier.quorum_voting_power()
+        {
+            let recheck = validator_verifier.check_voting_power(self.aggregated_signature.keys());
```

Comment:
Can you clarify why the recheck is needed?When is the  discrepancy expected?

Reply:
never. that's why I have error!() log.

Reply:
Hmm should we panic! instead in that case?